ddev/ddev
# fix: pass all valid options to nested commands in `ddev composer create`, fixes #6300, fixes #6246, for #5058

## The Issue

- #6300
- #6246
- #5058

Composer options were not respected by `composer run-script`.

## How This PR Solves The Issue

Makes `ddev composer create` fully compatible with all possible composer options. I removed all the hard-coded options.
Adds composer options to `composer run-script`.
Adds `--no-interaction` to CakePHP quickstart.

---

`--preserve-flags` (which doesn't add `--no-plugins` and `--no-scripts`) didn't work as expected, I decided to remove it.

This is not a breaking change, because `--preserve-flags` didn't work before (it was always `false`).

---

I moved `composer run-script post-create-project-cmd` outside the condition for `composer install`, to make it work the same way as in `composer create-project`.

---

If you passed `--no-scripts`, DDEV tried to run `composer run-script`, which was wrong behavior, fixed it.

---

If `--help` or `--version` flags were provided, composer showed it's help page or it's version, and DDEV tried to continue creating the project, which was wrong behavior, fixed it.

---

I refactored the test logic to make it more clear and added more tests. I removed the testing for the `psr/log` package and replaced it with custom-made fake packages.

## Manual Testing Instructions

1. The order of commands executed for `ddev composer create` must be the same as for `composer create-project`.

To check this behavior, you can use a test repository (it is used for `testdata` here) which contains all the necessary situations to check:

```
cd ~/workspace && mkdir composer-create && cd composer-create
ddev config --auto
mkdir -p .ddev/test-ddev-composer-create/{test-ddev-require,test-ddev-require-dev}

curl -o .ddev/test-ddev-composer-create/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/composer.json
curl -o .ddev/test-ddev-composer-create/test-ddev-require/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/test-ddev-require/composer.json
curl -o .ddev/test-ddev-composer-create/test-ddev-require-dev/composer.json https://raw.githubusercontent.com/stasadev/ddev/20240612_stasadev_composer_create/cmd/ddev/cmd/testdata/TestComposerCreateCmd/.ddev/test-ddev-composer-create/test-ddev-require-dev/composer.json

export repo='{"type": "path", "url": ".ddev/test-ddev-composer-create", "options": {"symlink": false}}'

# see the result for:
composer create-project --repository $repo test/ddev-composer-create # add different flags here

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create # add different flags here

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
````

`test/ddev-composer-create` package (`.ddev/test-ddev-composer-create/composer.json`) contains this:
- it requires `test/ddev-require` from `require` section:
  - `vendor/test/ddev-require` will be here, but only if there is no `--no-install` flag
- it requires `test/ddev-require-dev` from `require-dev` section:
  - `vendor/test/ddev-require-dev` will be here, but only if there is no `--no-dev`/`--no-install` flag
- it runs `post-root-package-install` script:
  - this script should run before `composer install`
  - it creates `created-by-post-root-package-install` file, but only if there is no `--no-scripts` flag
- it runs `post-create-project-cmd` script:
  - this script should run after `composer install`
  - it creates `created-by-post-create-project-cmd` file, but only if there is no `--no-scripts` flag

And use different flag combinations to see how it works for `composer create-project` and `ddev composer create`. It is important to check if the commands run in the same order.

Available flags for `composer create-project` https://getcomposer.org/doc/03-cli.md#create-project

Ignore the `test-ddev-require` and `test-ddev-require-dev` directories created in the project root (they are expected to be created every time as they are part of the composer package).

Every `ddev composer create` will have this output:
```
Executing composer command: [composer ...]
```
You can see what flags are used for each of these composer commands, and flags like `--no-interaction` will be also used for `composer run-script` and `composer install`.

For example:

This should not create `vendor` folder, and there should be no `created-by-*` files
```
# see the result for:
composer create-project --repository $repo test/ddev-composer-create --no-install --no-scripts

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create --no-install --no-scripts

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
```

This should create `vendor/test/ddev-require` folder, (but not `vendor/test/ddev-require-dev`) and there should be no `created-by-*` files
```
# see the result for:
composer create-project --repository $repo test/ddev-composer-create --no-scripts --no-dev

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf

# see the result for:
ddev composer create --repository $repo test/ddev-composer-create --no-scripts --no-dev

# clear the directory
ls | grep -xv ".ddev" | xargs rm -rf
```

And so on.

---

2. Try CakePHP quickstart (without interaction), it should not ask for input:

```
mkdir my-cakephp-site && cd my-cakephp-site
ddev config --project-type=cakephp --docroot=webroot
ddev composer create --prefer-dist --no-interaction cakephp/app:~5.0
```

---

3. Try `Steps To Reproduce` example (it should not ask for input) from:

- #6246

## Automated Testing Overview



## Release/Deployment Notes
